zookeeper-server/src/main/java/org/apache/zookeeper/server/ServerCnxn.java

```diff
@@ -622,4 +622,21 @@ public String getHostAddress() {
     public String getSessionIdHex() {
         return "0x" + Long.toHexString(getSessionId());
     }
+
+    /**
+     * Disposes the ZooKeeperSaslServer instance, if any, and any
+     * system resources or security-sensitive information it may be
+     * using.  The field is also set to <code>null</code>.  This
+     * method is idempotent.
+     */
+    protected final void disposeSaslServer() {
+        if (zooKeeperSaslServer != null) {
+            try {
+                zooKeeperSaslServer.dispose();
+            } catch (Exception e) {
```

Comment:
Is is possiblento catch a more specific list of exceptions?

Reply:
Fine with me; I just cargo-culted the block wrapping `sk.cancel()` in `NIOServerCnxn.close()`. (It means that the `SaslServer` will not be disposed of in the case of a `RuntimeException`, though.) Anyway: done.